File: node/chapt5/TF_ws/src/demo_python_tf/demo_python_tf/tf_listener.py
import rclpy
from rclpy.node import Node
import rclpy.time 
from tf2_ros import TransformListener,Buffer            # 坐标监听器，缓冲区
from tf_transformations import euler_from_quaternion    # 四元数 转 欧拉角 函数
import math                                             # 需要用到 math库 中的 角度转弧度的 函数

class TFListener(Node):      # 继承 Node类
    # pass
    def __init__(self):
        super().__init__('Dynamic_tf_broadcaster')        # 继承 父类 init函数，用来给 节点 其名字
        
        self.buffer_ = Buffer()

        self.listener_ = TransformListener(self.buffer_,self)         # 内部 需要创建 监听这 接收信息，需要Node=self
        
        # 创建 定时器 对象,并使用 定时器 定时 获取 tf 变换
        self.timer_ = self.create_timer(1,self.tf_transform)

    def tf_transform(self):
        '''
        功能：定时 获取 坐标关系（实时 查询 坐标关系）
        实现：利用 buffer 查询 坐标关系
        函数：
            lookup_transform('父坐标','子坐标',time,duration)：查询 父坐标 到 子坐标的 TF关系
            参数：       time：     查询的时刻。一般为 当前0时刻：rclpy.time.Time(seconds=0.0)      【Time类】
                        durantion：超时 期间。一般为 1s内的区间：rclpy.time.Duration(seconds=1.0)  【Duration类】

            euler_from_quaternion([x,y,z,w])：将 数组[x,y,z,w] 四元数 转换为 欧拉角(弧度值) 形式
            参数：      [x,y,z,w]：     为 transform.rotation.x/y/z/w
        '''
        try:
            # 1 . 将 查询变换的结果 放入 result中 ： 查询 base_link 到 bottole_link 的 第0时刻（当前时刻）的 坐标变换。若 超时1s，则 退出。
            result = self.buffer_.lookup_transform('base_link','bottle_link',
                                                   rclpy.time.Time(seconds=0.0),
                                                   rclpy.time.Duration(seconds=1.0))
            # 2. 若 查询成功，则 获取 TF 的 内容
            transform = result.transform
            self.get_logger().info(f'平移：{transform.translation}')
            self.get_logger().info(f'旋转：{transform.rotation}')           # 旋转 的 四元数 形式
            # 将 四元数 转换成 欧拉角 形式(传入的 需要是一个 数组格式)
            rotation_eular = euler_from_quaternion(
               [transform.rotation.x,
                transform.rotation.y,
                transform.rotation.z,
                transform.rotation.w
                ])
            
            self.get_logger().info(f'旋转RPY：{rotation_eular}')        # 旋转 的 欧拉角 形式

        # 异常 触发
        except Exception as e:
            self.get_logger().warn(f'获取坐标变换失败，原因{str(e)}')



# main函数
def main():
    # 1. ros初始化
    rclpy.init()
    # 2. 创建 节点类
    node = TFListener()
    # 3. 运行 节点
    rclpy.spin(node)
    # 4. 关闭 节点
    rclpy.shutdown()

File: node/chapt5/TF_ws/src/demo_cpp_tf/src/tf_listener.cpp
#include"rclcpp/rclcpp.hpp"
#include"geometry_msgs/msg/transform_stamped.hpp"
#include"tf2/LinearMath/Quaternion.hpp"                  // 提供 四元数 tf2::Quaternion类
#include"tf2_geometry_msgs/tf2_geometry_msgs.hpp"        // 消息类型 转换函数： 将 内容 转换成 消息接口
#include"tf2_ros/transform_listener.h"                   // 坐标 监听器 类
#include "tf2_ros/buffer.h"                              // 提供 buffer 类 的 库

#include "tf2/utils.h"                                   // 提供了 四元数据 转成 欧拉角 的 库

#include "chrono"
using namespace std::chrono_literals;

class TFListener:public rclcpp::Node
{
private:

    // 声明 监听器 对象
    std::shared_ptr<tf2_ros::TransformListener> listener_;

    // 声明 buffer 对象
    std::shared_ptr<tf2_ros::Buffer> buffer_;

    // 声明 定时器 对象
    rclcpp::TimerBase::SharedPtr timer_; 

public:
    // 构造函数 编写:对 声明的 对象 进行 实例化
    TFListener():Node("tf_listener")
    {   
        // 实例化 buffer 对象：用来 存储TF数据，在 回调函数中 查询
        buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());         // 需要 时间的记录
        // 实例化 listener，用来 获取/订阅 tf2_ros话题 的 数据，并将 其 放入 buffer 中
        this -> listener_ = std::make_shared<tf2_ros::TransformListener>(*buffer_,this);      // 使用 指针
        // 使用 定时器 不停 地 调用 函数 获取 TF坐标
        timer_ = this->create_wall_timer(1s,std::bind(&TFListener::get_transform,this)); // 使用 函数包装器 写 定时器
        
    }

    /**
     * 编写 获取 TF坐标 的方法
     * 方法：
     *      1. 
     *      2. 
     */
    void get_transform(){
        /**
         * 需要 捕获 异常： 使用 C++中的 try-catch 语句，进行 异常捕获
         */
        // 1. 在 buffer 里 查询 坐标关系
        try
        {
            // 查询 坐标关系：超时 1s 中 
            const auto transfer = buffer_ -> lookupTransform(
                "base_link",
                "target_point",
                this->get_clock()->now(),
                rclcpp::Duration::from_seconds(1.0f)        
            ); 
            // 2. 若 成功查询，则 执行 获取 查询transfer结果 存入 translation与rotation
            auto translation = transfer.transform.translation;
            auto rotation = transfer.transform.rotation;            // rotation中 包含 四元数

            // 3. 将 rotation中的 四元数 转化成 欧拉角
            // (1) 先 创建 Y P R 变量，用于 存入 Y P R 的 欧拉角
            double y,p,r;
            // (2) 使用tf2::getEulerYPR(rotation,y,p,r)方法：将 YPRW四元数 转化成 欧拉角,放入 y,p,r 变量中。
            tf2::getEulerYPR(rotation,y,p,r);                       // 参数 为 引用，直接更改 y,p,r
            
            // 4. 打印 translation与rotation 数据
            RCLCPP_INFO(get_logger(),"平移：%f,%f,%f",translation.x,translation.y,translation.z);
            RCLCPP_INFO(get_logger(),"旋转:%f,%f,%f",r,p,y);
        }
        catch(const std::exception& e)
        {
            // 若 查询失败，则 执行 报错
            RCLCPP_WARN(get_logger(),"%s",e.what());
        }
        
        
    }
};

int main(int argc,char *argv[]){
    rclcpp::init(argc,argv);

    auto node = std::make_shared<TFListener>();

    rclcpp::spin(node);

    rclcpp::shutdown();

    return 0;

}


